pennylane/measurements/mid_measure.py

```diff
@@ -99,68 +105,131 @@ class MeasurementValue(Generic[T]):
-        zero_case (float): the first measurement outcome value
-        one_case (float): the second measurement outcome value
+        measurement_ids (list of str): The id of the measurement that this object depends on.
+        fn (callable): A transformation applied to the measurements.
     """
 
-    __slots__ = ("_depends_on", "_zero_case", "_one_case", "_control_value")
-
-    def __init__(
-        self,
-        measurement_id: str,
-        zero_case: float = 0,
-        one_case: float = 1,
-    ):
-        self._depends_on = measurement_id
-        self._zero_case = zero_case
-        self._one_case = one_case
-        self._control_value = one_case  # By default, control on the one case
+    def __init__(self, measurement_ids, fn):
+        self.measurement_ids = measurement_ids
+        self.fn = fn
 
-    @property
-    def branches(self):
-        """A dictionary representing all the possible outcomes of the MeasurementValue."""
-        branch_dict = {}
-        branch_dict[(0,)] = self._zero_case
-        branch_dict[(1,)] = self._one_case
-        return branch_dict
+    def _items(self):
+        """A generator representing all the possible outcomes of the MeasurementValue."""
+        for i in range(2 ** len(self.measurement_ids)):
+            branch = tuple(int(b) for b in np.binary_repr(i, width=len(self.measurement_ids)))
+            yield branch, self.fn(*branch)
 
     def __invert__(self):
         """Return a copy of the measurement value with an inverted control
         value."""
-        inverted_self = copy.copy(self)
-        zero = self._zero_case
-        one = self._one_case
-
-        inverted_self._control_value = one if self._control_value == zero else zero
-
-        return inverted_self
-
-    def __eq__(self, control_value):
-        """Allow asserting measurement values."""
-        measurement_outcomes = {self._zero_case, self._one_case}
-
-        if not isinstance(control_value, tuple(type(val) for val in measurement_outcomes)):
-            raise MeasurementValueError(
-                "The equality operator is used to assert measurement outcomes, but got a value "
-                + f"with type {type(control_value)}."
+        return self._apply(lambda v: not v)
+
+    def __eq__(self, other):
+        if isinstance(other, MeasurementValue):
+            return self._merge(other)._apply(lambda v: v[0] == v[1])
+        return self._apply(lambda v: v == other)
+
+    def __add__(self, other):
+        if isinstance(other, MeasurementValue):
+            return self._merge(other)._apply(lambda v: v[0] + v[1])
+        return self._apply(lambda v: v + other)
+
+    def __radd__(self, other):
+        return self._apply(lambda v: other + v)
+
+    def __sub__(self, other):
+        if isinstance(other, MeasurementValue):
+            return self._merge(other)._apply(lambda v: v[0] - v[1])
+        return self._apply(lambda v: v - other)
```

Comment:
I wonder if the "structure" of such functions should be captured into a generic helper and then we could pass in two lambdas in each specific dunder method to make the methods more maintainable and readable.

```py
    def _apply_binary_transformation(self, fun_measurement_value, fun_general):
        if isinstance(other, MeasurementValue):
            return self._merge(other)._apply(fun_measurement_value)
        return self._apply(fun_general)

    def __sub__(self, other):
        return _apply_binary_transformation(lambda v: v[0] - v[1], lambda v: v - other)
```
I think this would make things simpler in most cases.

Reply:
yes, I considered this, my only thought was it's not too much additional code, but in some ways makes debugging easier since no use of lambdas? But I'm fine with either way

Reply:
I ended up doing this